### DISCUSSION TITLE
ESP32 on battery with ULP module - max deep sleep time and intended sensor-read mqtt delay for BME680

### DISCUSSION THREAD

**Discussion by belveder79** [arendst/Tasmota#21704] [Q&A]
I'm wondering about the setting of the ULP in deep sleep mode. I'm using this script which I borrowed from somewhere (don't remember).

```
import ULP
def goBackToSleep()
   ULP.wake_period(0,20000)
   var c = bytes().fromb64("dWxwAAwAzAAAABgAgwOAcg8AANA8AIBwEAAJggkBuC4wAMBwKAAAgAkB+C8PASByMADAcBAAQHIzA4ByDwAA0DMAAHAfAEByVABAgFMDgHJCA4ByDwAA0AsAAGgAAACwQwOAcg4AANAKAABycABAgBoAIHIOAABoAAAAsFMDgHJCA4ByDwAA0AsAAGgzA4ByDgAA0BoAAHIaAEByDgAAaGMDgHIOAADQGgAAcg4AAGhzA4ByDwAA0C8AIHC4AECAAAAAsDAAzCkQAEByuABAgAEAAJAAAACw")
   ULP.load(c)
   c = nil
   ULP.run()  
   print('Going back to sleep!')
   ULP.sleep(900)
end

print('Hello! Taking measurements...')
tasmota.set_timer(30000, goBackToSleep)
```

It runs on an ESP32 with a battery shield. The aim is to wake up periodically every 15 minutes to take a measurement of an attached I2C BME680 sensor and then go back to deep sleep. The BME680 is actually powered by an "Output-High" GPIO, so it is also off during deep sleep.

This works very well except for two things 
1) the BME680 needs some 10 seconds or so "swing in", so the humidity is usually at least 2% or more off from the more realistic value. Unfortunately, this value is sent immediately upon connection. Is there a delay which I can add to the standard sensor reading/mqtt sending in Tasmota to give it a little time to do a proper measurement? Alternatively, could I turn the auto-sending off and write a Berry function which does ```tasmota.read_sensors```and ```mqtt.publish``` in my own script (after some 10-15 seconds swing-in waiting)?

2) the maximum ULP.sleep(XXX) is 1800, so 30 minutes - for whatever reason I can't set it to anything higher. I have another system for a "surveilled" keybox, which would be fine to ping me once a day or so. The documentation does not say anything about the maximum sleep time, so here's the question: a) is this a Tasmota thing or in the ExpressIF framework or the chip itself? b) can it be changed and if yes, how?

Thx

**Comment by belveder79** [comment]
> the BME680 needs some 10 seconds or so "swing in", so the humidity is usually at least 2% or more off from the more realistic value. Unfortunately, this value is sent immediately upon connection. Is there a delay which I can add to the standard sensor reading/mqtt sending in Tasmota to give it a little time to do a proper measurement? Alternatively, could I turn the auto-sending off and write a Berry function which does tasmota.read_sensorsand mqtt.publish in my own script (after some 10-15 seconds swing-in waiting)?

I fixed the first issue now using ```TelePeriod 0``` and a simple function with a timer:
```
import mqtt
def sendValues()
  var s = tasmota.read_sensors()
  mqtt.publish('RESULT', s)
end
tasmota.set_timer(25000,sendValues)
```
Any "nicer" solution available?

**Reply by sfromis** [reply]
Your sending can be a oneliner:
`tasmota.set_timer(25000,/->mqtt.publish('RESULT', tasmota.read_sensors()))`

Powering a not-very-low-power sensor like BME680 may be borderline for a gpio, as that piece of rusty tin needs a "hotplate" to become functional for producing resistance values, still without clear relation to what's in the air.
